api/src/main/java/org/apache/iceberg/metrics/DefaultHistogram.java

```diff
@@ -0,0 +1,145 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.metrics;
+
+import java.util.Arrays;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+
+/**
+ * A default {@link Histogram} implementation that uses a fixed size reservoir like circular buffer.
+ * Oldest observation is replaced by latest observation.
+ */
+public class DefaultHistogram implements Histogram {
+  private final long[] measurements;
+  private long count;
+
+  public DefaultHistogram(int reservoirSize) {
+    this.measurements = new long[reservoirSize];
+    this.count = 0L;
+  }
+
+  @Override
+  public synchronized long count() {
+    return count;
+  }
+
+  @Override
+  public synchronized void update(long value) {
+    int index = (int) (count % measurements.length);
+    measurements[index] = value;
+    count++;
+  }
+
+  @Override
+  public Statistics statistics() {
+    int size = measurements.length;
+    if (count < (long) measurements.length) {
```

Comment:
what happens if a different thread calls `update()`? should we add a multi-threaded test similar to `TestDefaultTimer#multiThreadedStarts()` maybe?

Reply:
`update` is a `synchronized` method. 

the array clone is protected by `synchronized` too when reading array element
```
    long[] values = new long[size];
    for (int i = 0; i < values.length; ++i) {
      synchronized (this) {
        values[i] = measurements[i];
      }
    }
```

Reply:
Regarding multiple-thread testing, I will add similar test method.